Ì¥---
title: "RNA-seq FRDA Analysis Pipeline"
author: "Ankit Mukherjee"
output: html_document
---

# Load Libraries
```{r}
library(DESeq2)
library(limma)
library(tidyverse)
library(readxl)
library(writexl)
library(ggplot2)
library(BiocParallel)
library(patchwork)
library(ggrepel)
```

# Load RDS Files
```{r}
rds_files <- list.files(path = "salmon/", pattern = "\\.rds$", full.names = TRUE)
rds_list <- lapply(rds_files, readRDS)
names(rds_list) <- basename(rds_files)
```

# Load & Prepare Metadata
```{r}
attributes_frda <- read_xlsx("Metadata_RNAseq_FAvsHC.xlsx") |> subset(select = -Sno)
severity_scoring_frda <- read_xlsx("severity scoring.xlsx")
metadata_frda <- merge(attributes_frda, severity_scoring_frda, by = "Raw Data ID", all.x = TRUE)

metadata_frda$Batch <- factor(as.numeric(metadata_frda$Batch))
metadata_frda$Sex <- factor(metadata_frda$Sex)
metadata_frda$Age_scaled <- scale(as.numeric(metadata_frda$Age))
metadata_frda$Condition <- factor(metadata_frda$Condition, levels = c("Control", "Patient"))
metadata_frda$FSA_scaled <- scale(as.numeric(metadata_frda$`FSA scale`))
metadata_frda$mFARS_scaled <- scale(as.numeric(metadata_frda$`mFARS total`))
metadata_frda$mFARS_USS_scaled <- scale(as.numeric(metadata_frda$`mFARS USS`))
metadata_frda$DD_scaled <- scale(as.numeric(metadata_frda$DD))
metadata_frda$Onset_scaled <- scale(as.numeric(metadata_frda$Onset))
metadata_frda$HCM <- factor(as.numeric(metadata_frda$HCM), levels = c(0, 1))
metadata_frda$Diabetes <- factor(as.numeric(metadata_frda$Diabetes), levels = c(0, 1))
metadata_frda$GAA1_scaled <- scale(as.numeric(metadata_frda$GAA1))
metadata_frda$GAA2_scaled <- scale(as.numeric(metadata_frda$GAA2))

rownames(metadata_frda) <- metadata_frda$`Raw Data ID`
metadata_frda <- subset(metadata_frda, select = -`Raw Data ID`)
metadata_frda[metadata_frda == "NA"] <- NA

# Matching the rownames of all except the first to the first
all(sapply(rds_list[-1], function(x) identical(rownames(assay(x)), rownames(assay(rds_list[[1]])))))
```

# Clean Sample Sets
```{r}
# Removing batch3 top up samples from batch3
samples_to_remove <- intersect(colnames(assay(rds_list[[3]])), colnames(assay(rds_list[[4]])))
rds_list[[3]] <- rds_list[[3]][, !colnames(assay(rds_list[[3]])) %in% samples_to_remove]

# Remove PC samples in Batch8
rds_list[[9]] <- rds_list[[9]][, !colnames(assay(rds_list[[9]])) %in% c("PC1", "PC2", "PC3")]
```

# Combine Count Data
```{r}
combined_rds <- do.call(cbind, rds_list)
count_data <- assay(combined_rds)
count_data <- count_data[, sort(colnames(count_data))]

# Replace Ensembl IDs with gene names
gene_names <- rowData(rds_list[[1]])$gene_name
rownames(count_data) <- make.unique(gene_names)

# Filter low-expression genes
dim(count_data)
count_data_filtered <- count_data[rowSums(count_data >= 10) >= 5, ]
dim(count_data_filtered)
```

# Check Design Matrix
```{r}
# Check whether the model matrix is full rank or not 
# (i.e. there are no collinearities between covariates in the design formula)
check_design_collinearity <- function(formula, data) {
  design_matrix <- model.matrix(formula, data = data)
  # Check rank
  is_full_rank <- qr(design_matrix)$rank == ncol(design_matrix) # QR decomposition to check rank
  cat("Design matrix full rank: ", is_full_rank, "\n")
  if (!is_full_rank) {
    cat("\nDetected linear dependencies:\n")
    lm_obj <- lm(rep(1, nrow(design_matrix)) ~ design_matrix) # Fit linear model to identify dependencies
    print(alias(lm_obj)) # Display linear dependencies
  } else {
    cat("No linear dependencies detected.\n")
  }
  invisible(design_matrix)
}
```

# Differential expression between patients and controls from all batches
```{r}
all.equal(rownames(metadata_frda), colnames(count_data_filtered))
# dds <- DESeqDataSetFromMatrix(countData = round(count_data_filtered), 
#                               colData = metadata_frda, 
#                               design = ~ Batch + Sex + Age_scaled + Condition)
# dds <- estimateSizeFactors(dds)
# dds <- DESeq(dds, parallel = TRUE)
# saveRDS(dds, "dds_frda.rds")
dds <- readRDS("dds_frda.rds")
```

# Batch Correction
```{r}
vsd <- vst(dds, blind = FALSE)
expr_vst <- assay(vsd)
covs <- model.matrix(~ Sex + Age + Condition, data = metadata_frda)
expr_rb <- limma::removeBatchEffect(expr_vst, batch = metadata_frda$Batch, design = covs)
```

# PCA Plot Function
```{r}
# PCA plot before and after removing batch effects with top 500 variable genes
# plotPCA(vsd, intgroup = c("Batch", "Condition")) # before batch effect removal
# or
plot_pca <- function(expr_matrix, metadata, top_n = 500, color_by, shape_by, title) {
  metadata <- metadata[colnames(expr_matrix), , drop = FALSE]
  topVarGenes <- order(matrixStats::rowVars(expr_matrix), decreasing = TRUE)[seq_len(top_n)]
  
  pca_res <- prcomp(t(expr_matrix[topVarGenes, ]))
  percent_var <- pca_res$sdev^2 / sum(pca_res$sdev^2)
  
  pca_data <- data.frame(
    Sample = colnames(expr_matrix),
    PC1 = pca_res$x[, 1],
    PC2 = pca_res$x[, 2],
    Color = metadata[[color_by]]
  )
  
  if (!is.null(shape_by)) {
    pca_data$Shape <- metadata[[shape_by]]
  }
  
  pca_data <- na.omit(pca_data)
  
  p <- ggplot(pca_data, aes(x = PC1, y = PC2)) +
    geom_point(
      aes(color = Color, shape = if (!is.null(shape_by)) Shape else NULL),
      size = 3, alpha = 0.8
    ) +
    theme_bw(base_size = 14) +
    labs(
      title = title,
      x = paste0("PC1: ", round(percent_var[1] * 100, 1), "% variance"),
      y = paste0("PC2: ", round(percent_var[2] * 100, 1), "% variance"),
      color = color_by,
      shape = shape_by
    ) +
    theme(
      plot.title = element_text(hjust = 0.5, face = "bold"),
      legend.position = "right"
    )
  
  return(p)
}
```

# DGE
```{r}
# coef from resultsNames(dds)
# res_contrast <- results(dds, contrast = c("Condition", "Patient", "Control")) |> 
#   as.data.frame() |> 
#   rownames_to_column("gene_names")

# res <- lfcShrink(dds, coef = "Condition_Patient_vs_Control", type = "apeglm", parallel = TRUE) |> 
#   as.data.frame() |> 
#   rownames_to_column("gene_names")
# saveRDS(res, "res_frda.rds")
res <- readRDS("res_frda.rds")
```

# Volcano Plot Function
```{r}
volcano_plot <- function(res,
                         gene_label_col,
                         lfc_threshold = 2,
                         padj_threshold = 0.01,
                         top_n = 10,
                         base_size = 14) {

  vol_data <- res |>
    na.omit() |> 
    mutate(Expression = case_when(
      log2FoldChange >= lfc_threshold & padj <= padj_threshold ~ "Upregulated",
      log2FoldChange <= -lfc_threshold & padj <= padj_threshold ~ "Downregulated",
      TRUE ~ "Not significant"
    ))

  top_up <- vol_data |>
    filter(Expression == "Upregulated") |>
    arrange(padj) |>
    head(top_n)
  
  top_down <- vol_data |>
    filter(Expression == "Downregulated") |>
    arrange(padj) |>
    head(top_n)
  
  top_genes <- bind_rows(top_up, top_down)

  p <- ggplot(vol_data, aes(x = log2FoldChange, y = -log10(padj))) +
    geom_point(aes(color = Expression)) +
    geom_text_repel(
      data = top_genes,
      aes(label = .data[[gene_label_col]]),
      size = 4,
      max.overlaps = 15
    ) +
    scale_color_manual(values = c(
      "Upregulated" = "#D55E00",
      "Downregulated" = "#0072B2",
      "Not significant" = "gray70"
    )) +
    scale_x_continuous(name = "Log2 fold change") +
    scale_y_continuous(name = "-Log10 adjusted p-value") +
    geom_hline(yintercept = -log10(padj_threshold), linetype = 2, color = "black") +
    geom_vline(xintercept = c(-lfc_threshold, lfc_threshold), linetype = 2, color = "black") +
    theme_classic(base_size = base_size) +
    theme(
      panel.grid = element_blank(),
      plot.background = element_rect(fill = "white"),
      legend.position = "right"
    )
  
  return(p)
}
```

# Comparing FXN expression levels across patients and controls
```{r, eval=FALSE}
fxn_rb <- expr_rb["FXN", ]
fxn_df_pre <- count_data["FXN", ] |> 
              rownames_to_column("Gene") |> 
              pivot_longer(cols = -Gene, names_to = "SampleID", values_to = "fxn_pre") |> 
              select(-Gene)
fxn_df_post <- fxn_rb |> as.data.frame() |> rownames_to_column("SampleID")
write_xlsx(fxn_df_pre, "fxn_df_pre.xlsx")
write_xlsx(fxn_df_post, "fxn_exp.xlsx")

data_fxn <- metadata_frda
data_fxn$FXN_expression <- fxn_rb
data_fxn <- data_fxn[order(data_fxn$FXN_expression), ]
data_fxn <- data_fxn |> rownames_to_column("SampleID")
highlight_samples <- c("FA18", "FA25", "FA43", "FA45")
data_fxn$highlight <- ifelse(data_fxn$SampleID %in% highlight_samples, "Highlight", "Other")

fxn_exp <- ggplot(data_fxn, aes(x = reorder(SampleID, FXN_expression),
                      y = FXN_expression)) +
  geom_point(aes(color = Condition, alpha = highlight, size = highlight)) +
  scale_alpha_manual(values = c("Other" = 0.5, "Highlight" = 1)) +
  scale_size_manual(values = c("Other" = 2, "Highlight" = 4)) +
  geom_text(data = subset(data_fxn, highlight == "Highlight"),
            aes(label = SampleID),
            vjust = -1.0, size = 3, color = "black", fontface = "bold") +
  theme_bw(base_size = 14) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  labs(x = "Samples", y = "FXN expression")

ggsave("fxn_exp.jpeg",
        plot = fxn_exp, 
        device = "jpeg",
        units = "in",
        height = 8,
        width = 25,
        dpi = 600)
```

# Segregating samples on the basis of batch effects into test and replication cohort

```{r, eval=FALSE}
# Removing samples FA18, HC8, FA43, FA25, FA81, FA45 and FA82 (from FXN exp plot)
samples_to_remove_fxn_abnormal <- c("FA18", "HC8", "FA43", "FA25", "FA81", "FA45", "FA82")
metadata_frda_remove_fxn_abnormal <- metadata_frda[!rownames(metadata_frda) %in% samples_to_remove_fxn_abnormal, ]
count_data_remove_fxn_abnormal <- count_data_filtered[, rownames(metadata_frda_remove_fxn_abnormal)]

pca_data <- plotPCA(vsd, intgroup = c("Batch", "Condition"), returnData = TRUE)
test_batches <- pca_data$Batch[pca_data$PC1 < 0] |> unique()
replication_batches <- pca_data$Batch[pca_data$PC1 > 0] |> unique()

test_metadata <- metadata_frda_remove_fxn_abnormal[metadata_frda_remove_fxn_abnormal$Batch %in% test_batches, ]
test_count_data <- count_data_filtered[, rownames(test_metadata)]
replication_metadata <- metadata_frda_remove_fxn_abnormal[metadata_frda_remove_fxn_abnormal$Batch %in% replication_batches, ]
replication_count_data <- count_data_filtered[, rownames(replication_metadata)]
```

# DGE function
```{r}
dge <- function(counts, meta, design, filename, filename_volcano) {
  library(DESeq2)
  library(writexl)
  library(ggplot2)

  dds <- DESeqDataSetFromMatrix(countData = round(counts),
                                colData = meta,
                                design = design)
  dds <- DESeq(dds, parallel = FALSE)
  
  cat("\nCoefficients:\n")
  print(resultsNames(dds))

  mode_choice <- readline(prompt = "\nUse 'coef' or 'contrast'? ")
  
  if (tolower(mode_choice) == "coef") {
    coef_choice <- readline(prompt = "Enter the coefficient name: ")
    res <- lfcShrink(dds, coef = coef_choice, type = "apeglm") |>
      as.data.frame() |>
      rownames_to_column("gene_names")
    
  } else if (tolower(mode_choice) == "contrast") {
    cat("\nEnter contrast as a single R expression.\n")
    cat("Examples:\n")
    cat("# Biological Question                               | Intercept Model (~ genotype + cond + g:c)                    | No-Intercept Model (~ 0 + group)\n")
    cat("# --------------------------------------------------------------------------------------------------------------------------\n")
    cat("# Condition effect in Genotype I (B vs A in I)       | results(dds, name=\"condition_B_vs_A\")                     | results(dds, contrast=list(\"groupI_B\", \"groupI_A\"))\n")
    cat("# Condition effect in Genotype II (B vs A in II)     | results(dds, list(c(\"condition_B_vs_A\", \"genotypeII.conditionB\"))) | results(dds, contrast=list(\"groupII_B\", \"groupII_A\"))\n")
    cat("# Interaction effect (Is B vs A different in II vs I?) | results(dds, name=\"genotypeII.conditionB\")               | results(dds, contrast=list(c(\"groupII_B\", \"groupI_A\"), c(\"groupII_A\", \"groupI_B\")))\n")
    cat("# ---------------------------------------------------------------------------\n")
    cat("Wrap the contrast in list(), e.g. list(\"Condition_Treated_vs_Control\") or list(\"Batch1.ConditionControl\", \"Batch9.ConditionControl\")\n\n")
    
    contrast_input <- readline(prompt = "Enter your contrast: ")
    contrast_list <- eval(parse(text = contrast_input))
    
    res <- lfcShrink(dds, contrast = contrast_list, type = "ashr") |>
      as.data.frame() |>
      rownames_to_column("gene_names")
    
  } else {
    stop("Invalid input")
  }
  
  resOrdered <- res[order(res$padj), ]
  write_xlsx(resOrdered, filename)
  
  volcano <- volcano_plot(resOrdered, gene_label_col = "gene_names")
  ggsave(
    filename = filename_volcano,
    plot = volcano,
    device = "jpeg",
    units = "in",
    height = 8,
    width = 15,
    dpi = 600
  )
  
  return(resOrdered)
}
```

# Finding differences between patients of Batches 1 to 7 and Batches 8 and 9
```{r, eval=FALSE}
metadata_frda_remove_fxn_abnormal <- metadata_frda_remove_fxn_abnormal %>%
  mutate(Batch_group = ifelse(Batch %in% 1:7, "Batch 1-7", "Batch 8-9"),
         GAA1_GAA2 = GAA1 + GAA2) %>%
  filter(Condition == "Patient") %>%
  mutate(across(c(Age, DD, Onset, GAA1, GAA2, GAA1_GAA2), as.numeric))

plot_box_jitter_vars <- function(data, y_vars, group_col = "Batch_group") {
  
  plots <- lapply(y_vars, function(var) {
    ggplot(data, aes_string(x = group_col, y = var)) +
      geom_boxplot(outlier.shape = NA, fill = "lightblue") +
      geom_jitter(width = 0.2, size = 2, alpha = 0.7, color = "darkblue") +
      theme_bw(base_size = 14)
  })

  names(plots) <- y_vars
  return(plots)
}

vars_to_plot <- c("Age", "DD", "Onset", "GAA1", "GAA2", "GAA1_GAA2")
plots <- plot_box_jitter_vars(metadata_frda_remove_fxn_abnormal, vars_to_plot)
wrap_plots(plots, ncol = 3)

# Read alignment differences
reads_aligned <- read_excel("QC/QC/qc_result.xlsx", sheet = "millions_aligned")

p <- ggplot(reads_aligned, aes(x = Sample, y = `M Aligned`)) +
  geom_bar(stat = "identity", fill = "steelblue") +
  geom_text(aes(label = round(`M Aligned`, 1)), vjust = -0.5, size = 6) +
  facet_grid(. ~ Condition + Batch, scales = "free_x", space = "free_x") +
  scale_y_continuous(expand = c(0, 0)) +
  theme_bw(base_size = 18) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  labs(x = "Samples", y = "Millions aligned")

ggsave("QC/QC/reads_aligned.jpeg",
       plot = p,
       device = "jpeg",
       units = "in",
       height = 20,
       width = 50,
       dpi = 600,
       limitsize = FALSE)
```

# DE for batch 8 and 9 patients and controls
```{r, eval=FALSE}
# Selecting samples that have million mapped reads to the transcriptome > 5 
# and million mapped fragments to the exonic region of the genome > 10

qc_result <- read_excel("QC/qc_dragen_salmon.xlsx", skip = 1)
samples_qualified <- qc_result %>%
  filter(`M Aligned` > 5 & `Exonic` > 10) %>% # exonic info available for batch 8 and 9 only
  pull(Sample)

metadata_frda_batch8_9 <- metadata_frda %>%
  filter(Batch %in% c(8, 9) & rownames(metadata_frda) %in% samples_qualified)
metadata_frda_batch8_9$Batch <- factor(metadata_frda_batch8_9$Batch)
count_data_frda_batch8_9 <- count_data_filtered[, rownames(metadata_frda_batch8_9)]

batch8_9_result <- dge(counts = count_data_frda_batch8_9, 
                      meta = metadata_frda_batch8_9,
                      design = ~ Batch + Sex + Age_scaled + Condition,
                      filename = "res/res_batch8_9_patients_vs_controls.xlsx",
                      filename_volcano = "volcano_plot/volcano_batch8_9_patients_vs_controls.jpeg")
```